lib/Common/Memory/HeapAllocator.cpp

```diff
@@ -146,7 +153,15 @@ void HeapAllocator::Free(void * buffer, size_t byteSize)
         return;
     }
 #endif
-    free(buffer);
+
+    if (CONFIG_FLAG(PrivateHeap))
+    {
+        HeapFree(GetPrivateHeap(), 0, buffer);
```

Comment:
If we are going this direction, maybe we can have different serialized/non-serialized heap? per thread heap can have HEAP_NO_SERIALIZE flag set. Only limited arena need cross-thread access.